llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp

```diff
@@ -376,6 +389,124 @@ bool RISCVInstructionSelector::selectSExtInreg(MachineInstr &MI,
   return true;
 }
 
+/// Returns the RISCVCC::CondCode that corresponds to the CmpInst::Predicate CC.
+/// CC Must be an ICMP Predicate.
+static RISCVCC::CondCode getRISCVCCFromICMP(CmpInst::Predicate CC) {
+  switch (CC) {
+  default:
+    llvm_unreachable("Expected ICMP CmpInst::Predicate.");
+  case CmpInst::Predicate::ICMP_EQ:
+    return RISCVCC::COND_EQ;
+  case CmpInst::Predicate::ICMP_NE:
+    return RISCVCC::COND_NE;
+  case CmpInst::Predicate::ICMP_ULT:
+    return RISCVCC::COND_LTU;
+  case CmpInst::Predicate::ICMP_SLT:
+    return RISCVCC::COND_LT;
+  case CmpInst::Predicate::ICMP_UGE:
+    return RISCVCC::COND_GEU;
+  case CmpInst::Predicate::ICMP_SGE:
+    return RISCVCC::COND_GE;
+  }
+}
+
+MachineInstr *RISCVInstructionSelector::createICMPForBranch(
```

Comment:
Why do we need to build a new ICmp? Can't we just give the caller the registers it needs to pass to the select?

Reply:
I built a new ICmp because it packs everything that is needed into a single data structure. Are you suggesting to have `createICMPBranch` be called something like `getSelectOperandsFromICMP` which returns a custom data structure that isn't a MachineInstr?

Reply:
Yes. Or just pass 3 variables by reference like `translateSetCCForBranch` in SelectionDAG.

Creating a new MachineInstr has to heap allocate it, insert it into the basic block instruction linked list, update MachineRegisterInfo. Maybe some other things. That seems like overkill.

Reply:
Updated. You make a good point.